Remove parameters not needed after move to YAML-based publishing

JohnTortugo opened the pull request:
Closes: https://github.com/dotnet/arcade/issues/4604

- Assumes that PublishUsingPipelines will be true for all repos that want to publish assets.
- Remove things that are only used in the Release Pipelines - since we are going to delete them.
- Remove copy of assets/manifest to temporary folder. That's not needed anymore.

mmitche commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj:
```diff
@@ -144,10 +138,6 @@
     <ItemGroup>
       <_PublishProps Include="@(_CommonProps)"/>
-      <_PublishProps Include="PublishToAzureDevOpsArtifacts=$(DotNetPublishUsingPipelines)" />
-      <_PublishProps Include="ArtifactsCategory=$(DotNetArtifactsCategory)" />
```
This might still be used by dotnet/toolset. Their zip file gets published to the dotnet-toolset container in dotnetfeed.  This could be changed though, and then this could be removed.

JohnTortugo commented:
> Looks good so far. Question about the dotnet/toolset still stands (maybe I'm misunderstanding with the artifact category would feed through for them).

The `ArtifactsCategory` in those .proj are only used by the Release Pipeline Runner. Those .proj don't forward that variable to the post-build stages. Does this address your concern?

> What's your validation strategy for this?

The plan that I had was to go around in different repos and spin builds using a custom branch + a test SDK. Do you have a suggestion for how to better testing that?

JohnTortugo commented:
Can you all PTAL again?

JohnTortugo commented:
@tmat - Thanks for the comment. I think I've addressed that concern in my latest commit. I added a property called `PublishToSymbolServer` to Build.proj that should capture the scenarios where people want to publish to the symbol servers during the build.

I'm going to ran a few builds to test this in different scenarios. It looks like [NuGet.BuildTasks](https://github.com/dotnet/NuGet.BuildTasks/blob/master/.vsts-ci.yml#L32) and [dotnet/interactive](https://github.com/dotnet/interactive/blob/master/azure-pipelines.yml#L59) would be good test cases. Still, I'd appreciate if you could review again and please let me know which repos did you have in mind when you thought about this scenario.

tmat commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:
```diff
@@ -17,15 +15,12 @@
-    
+
     <PublishToSymbolServer>false</PublishToSymbolServer>
     <PublishToSymbolServer Condition="'$(UsingToolSymbolUploader)' == 'true' and '$(AzureFeedUrl)' == '' and '$(ContinuousIntegrationBuild)' == 'true'">true</PublishToSymbolServer>
```
>AzureFeedUrl [](start = 84, length = 12)

I guess AzureFeedUrl shouldn't be used here since you removed it from the comments above?

tmat replied:
I think we should have a single variable that determines whether we are publishing to BAR or not. 
If we are then no further publishing should be done.
If we are not then we publish to symbol servers.

Previously this was determined based on the value of AzureFeedUrl.

---
In reply to: [372078995](https://github.com/dotnet/arcade/pull/4626#discussion_r372078995) [](ancestors = 372078995)

tmat replied:
Yes, it does. What I'm suggesting though is that we don't pivot on symbol publishing. What I propose is to replace the property `PublishToSymbolServer` that's defined in Build.proj with `PublishToBuildAssetRegistry` (with negated condition). Then use `PublishToBuildAssetRegistry` to control whether we are publishing to BAR or whether we are publishing directly. When publishing directly we publish the symbols.

tmat replied:
Then the property `PublishToSymbolServer` defined in Publish.proj would be defined as `'$(PublishToBuildAssetRegistry)' != 'true' and '$(UsingToolSymbolUploader)' == 'true' and '$(ContinuousIntegrationBuild)' == 'true'`

tmat replied:
In addition, `PublishToAzureDevOpsArtifacts` would be conditional on `$(PublishToBuildAssetRegistry)' == 'true'`

JohnTortugo replied:
I got the idea now. I think that's a good plan. I only have one issue/question: some repos publish the symbols during the build stage because they have `UsingToolSymbolUploader == true` and `AzureFeedUrl == ""`. If we don't include `AzureFeedUrl` in the conditions then these repos will stop publishing symbols during the build stage. Are you thinking we should go to these repos and update then to set `PublishToBuildAssetRegistry == false` ?

JohnTortugo replied:
Talked offline.

tmat commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:
```diff
@@ -2,12 +2,10 @@
-      AzureFeedUrl                      Target Azure feed URL.
-      AzureAccountKey                   Azure account key.
       DotNetFinalVersionKind            Global property that stores the type of the current build: https://github.com/dotnet/arcade/blob/master/Documentation/CorePackages/Versioning.md#build-kind
       DotNetOutputBlobFeedDir           Source Build publishing directory
```
> DotNetOutputBlobFeedDir           Source Build publishing directory [](start = 5, length = 68)

Is Source Build still using this?

JohnTortugo replied:
Looks like they are. [dirs.targets](https://github.com/dotnet/source-build/blob/81b5f5cf7272810d44a5876838c79dbdd607cfee/repos/dir.targets#L57) and [core-setup.proj](https://github.com/dotnet/source-build/blob/81b5f5cf7272810d44a5876838c79dbdd607cfee/repos/core-setup.proj#L22). @dagood - do you know about this?

dagood replied:
Yep, it is.

tmat commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj:
```diff
@@ -89,6 +85,14 @@
+        we publish symbols during the build stage.
+      -->
+      <PublishToSymbolServer Condition="'$(DotNetPublishUsingPipelines)' != 'true'">true</PublishToSymbolServer>
+      <PublishToSymbolServer Condition="'$(DotNetPublishUsingPipelines)' == 'true'">false</PublishToSymbolServer>
```
Nit: can be simplified a bit:
```
<PublishToSymbolServer>true</PublishToSymbolServer>
<PublishToSymbolServer Condition="'$(DotNetPublishUsingPipelines)' == 'true'">false</PublishToSymbolServer>
```

JohnTortugo commented:
I talked offline with @tmat. Here is the new plan:

We are going to use the `DotNetPublishUsingPipelines` property as an decision variable like this:

- When `DotNetPublishUsingPipelines` == `true` we'll publish to BAR and let Maestro handle **all** publishing, including symbols.
- When `DotNetPublishUsingPipelines` == `false` we won't publish to BAR *but* symbol publishing will be performed immediately by Arcade SDK.
- `DotNetPublishUsingPipelines` doesn't have a default value. All repos publishing to BAR already set it so we are going to keep that.

I've patched the code already. I'll spin some test builds.

tmat commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj:
```diff
@@ -129,8 +133,13 @@
       <_RestoreToolsProps Include="_NuGetRestoreTargets=$(_NuGetRestoreTargets)"/>
+      
+      <!-- Used in the SDK (Tools.proj) to control whether Build.Tasks.Feed will be restored or not. -->
+      <_RestoreToolsProps Include="Publish=$(Publish)"/>
```
>Publish [](start = 45, length = 7)

Shouldn't this be DotNetPublishUsingPipelines? If we don't publish to BAR we don't need Build.Tasks.Feed, right?

JohnTortugo replied:
I think we can't. Because even though we don't publish to BAR people can still try to use tasks from Tasks.Feed in their `\eng\publishing.props`

JohnTortugo replied:
[Core-SDK is an example](https://github.com/dotnet/core-sdk/blob/master/eng/Publishing.props#L117), not sure if it's the only one.

tmat replied:
That looks like a workaround for something we are missing. I'd expect only the SDK calls this task and that the publishing.props file just adds to predefined list of items. Like we do for signing. For now, let's keep it, but I'd be good to follow up on this.

tmat commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:
```diff
@@ -37,19 +30,15 @@
   </PropertyGroup>
 
-  <Import Project="$(NuGetPackageRoot)microsoft.dotnet.build.tasks.feed\$(MicrosoftDotNetBuildTasksFeedVersion)\build\Microsoft.DotNet.Build.Tasks.Feed.targets" Condition="$(PublishToAzure)"/>
+  <Import Project="$(NuGetPackageRoot)microsoft.dotnet.build.tasks.feed\$(MicrosoftDotNetBuildTasksFeedVersion)\build\Microsoft.DotNet.Build.Tasks.Feed.targets"/>
```
>Microsoft.DotNet.Build.Tasks.Feed.targets [](start = 118, length = 41)

Shouldn't this be conditional on DotNetPublishUsingPipelines==true?

JohnTortugo commented:
@mmitche - PTAL when you have some spare cycles.

I've tested these changes in these repos and they all passed: dotnet/runtime, dotnet/winforms, dotnet/wpf, dotnet/sdk, dotnet/arcade, dotnet/wpf-int, dotnet/toolset, dotnet/templating, dotnet/efcore

I'm running new builds on all repos above to test some minor fixes I did after the previous tests:

- dotnet/runtime, 
- dotnet/winforms, 
- dotnet/wpf, 
- dotnet/sdk, 
- dotnet/arcade, 
- dotnet/wpf-int, 
- dotnet/toolset, 
- dotnet/templating, 
- dotnet/efcore

JohnTortugo commented:
Tests on repos that aren't using Arcade publishing infra:

- [VSDebugCore](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3438758&view=results)

mmitche commented:
Thanks for the tests. I'd run core-sdk if you can.

JohnTortugo commented:
Testing again in these repos after a in Arcade.SDK\DefaultVersions.props to make the latest Tasks.Feed the default version:

- [runtime](https://dnceng.visualstudio.com/internal/_build/results?buildId=506212&view=results)
- [aspnetcore](https://dnceng.visualstudio.com/internal/_build/results?buildId=506213&view=results)
- [core-sdk](https://dnceng.visualstudio.com/internal/_build/results?buildId=506210&view=results)
- [windowsdesktop](https://dnceng.visualstudio.com/internal/_build/results?buildId=506211&view=results)
- [vsdebugcore](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3439801&view=results)

riarenas commented:
Thanks for the tests. Could you add Roslyn to your test matrix so that we are sure this won't affect devdiv repos?

JohnTortugo commented:
- We had [VSDebugCore](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3439801&view=results) on DevDiv. 
- I started now a build for [Roslyn-Signed](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=3443173&view=results)

JohnTortugo commented:
Roslyn build also passed.

riarenas commented:
> We are going to use the DotNetPublishUsingPipelines property as an decision variable like this:

When DotNetPublishUsingPipelines == true we'll publish to BAR and let Maestro handle all publishing, including symbols.
When DotNetPublishUsingPipelines == false we won't publish to BAR but symbol publishing will be performed immediately by Arcade SDK.
When DotNetPublishUsingPipelines == "" we make it false.

This change in behavior should be added to the documentation

JohnTortugo commented:
@riarenas - Is this the only concern here? I can definitely include it in this PR but if it's the only concern perhaps I can address it in a separate PR.

riarenas commented:
No, I just got distracted by other things going on and that was the obvious thing that popped up.

riarenas commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:
```diff
@@ -2,12 +2,11 @@
     Optional variables:
-      AzureFeedUrl                      Target Azure feed URL.
-      AzureAccountKey                   Azure account key.
+      DotNetPublishUsingPipelines       Whether publishing to BAR should be performed or not. If set to 'false' symbol publishing will be performed immediately.
```
With the whole context, this comment makes sense, but it's probably too confusing for its own good.

If this controls whether we publish to BAR or not, what does symbol publishing have to do with it?

JohnTortugo replied:
Changed the words a bit.

riarenas commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:
```diff
@@ -2,12 +2,11 @@
       DotNetOutputBlobFeedDir           Source Build publishing directory
-      DotNetSymbolServerTokenMsdl       Personal access token for MSDL symbol server. Available from variable group DotNet-Symbol-Publish.
-      DotNetSymbolServerTokenSymWeb     Personal access token for SymWeb symbol server. Available from variable group DotNet-Symbol-Publish.
+      DotNetSymbolServerTokenMsdl       Personal access token for MSDL symbol server. Available from variable group DotNet-Symbol-Publish.	
```
[nit] Is the change here a trailing space?

riarenas commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:
```diff
@@ -140,71 +110,25 @@
-        Adding this just to not fail the validation when creating the manifest. 
-        TODO: Should be removed once PublishingUsingPipelines is the default.
-      -->
-      <NewManifestBuildData Include="Location=$(AzureFeedUrl)" />
```
Could you double check if we can actually remove this? I know we've discussed issues with this attribute in the past.

JohnTortugo replied:
Yeap. I'll post a comment below describing this.

riarenas commented on src/Microsoft.DotNet.Arcade.Sdk/tools/Publish.proj:
```diff
@@ -261,15 +176,15 @@
 
-      <!--
-        Publish Portable PDBs contained in symbol packages.
+      <!--	
```
All these just line ending changes?

riarenas commented on src/Microsoft.DotNet.Build.Tasks.Feed/src/BuildManifestUtil.cs:
```diff
@@ -139,14 +139,14 @@ public static void WriteAsXml(this BuildModel buildModel, string filePath, TaskL
+        internal static bool ManifestBuildDataHasLocationInformation(IDictionary<string, string> attributes)
         {
-            return attributes.ContainsKey("Location");
+            return attributes.ContainsKey("Location") || attributes.ContainsKey("InitialAssetsLocation");
```
Oh, is this what makes it fine to stop populating the location property?

JohnTortugo replied:
Yeap. I'll post a comment below describing this.

JohnTortugo commented:
@riarenas - Yeah, removing the "Location=xxx" attribute from the manifest here is backed up by two things: 1) The validation in Tasks.Feed is being patched to relax the condition and 2) the Publish to BAR task requires _either_ Location or InitialAssetsLocation attribute, we are keeping the latter.

A potential risk is if some repo has an old version for Tasks.Feed in eng\version.props (and not in eng\Version.Details.xml) - that version would include the validation that still requires the Location attribute. If we release the SDK created by this PR and it picks that old version of Tasks.Feed then the validation would fail. To reduce the chances of this happening I did:

- Included in this PR a change to Arcade SDK DefaultVersions.props that makes the default version of Tasks.Feed be the same as current SDK version.
- I did a github search on the dotnet/ and microsoft/ orgs and didn't find any repo that hardcode a version of Tasks.Feed in eng\version.props but not in eng\Version.Details.xml

So, I think we are safe dropping the attribute.
